File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
  const newNode = new ListNode(_, head);
  let currentNode = newNode;
  let totalNodeCount = 0;
  let count = 0;

  while (currentNode) {
    totalNodeCount++;
    currentNode = currentNode.next;
  }
 
  if (totalNodeCount === 1 && n === 1) return null;
  if (totalNodeCount - 1 === n) return head.next;

  currentNode = newNode;

  while (currentNode) {
    if (count === totalNodeCount - n - 1) {
      currentNode.next = currentNode?.next?.next ?? null;
      break;
    }

    count++;
    currentNode = currentNode.next;
  }

  return head;
};

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        cur_node = head
        total_length = 0
        
        while cur_node:
            cur_node = cur_node.next
            total_length += 1
        
        if total_length == n:
            return head.next if head.next else None
        
        cur_node = head
        target = total_length - n
        count = 0

        while cur_node.next:
            if count == target - 1:
                cur_node.next = cur_node.next.next
                return head

            cur_node = cur_node.next
            count += 1
        
        cur_node.next = None
        return head



                